### ISSUE TITLE
[SPMD]  Can mark_sharding operation automatically insert all_reduce or all_gather operators?

### ISSUE THREAD

**Issue by baoleai** [pytorch/xla#5180]
## ❓ Questions and Help
Hi, @JackCaoG @yeounoh , I am trying to use `mark_sharding` to implement Tensor Parallelism and would like to confirm whether the shard operation will automatically insert appropriate communication operators like all_reduce or all_gather. 
For example, if we have an MLP consisting of two linear layers A and B, where Z=Dropout(YB) and Y=GeLu(XA), we split A by column and B by row. After splitting, we need to add an all_reduce operator to A's backward and B's forward to make the result equivalent to before the splitting.  Is this automatically done by xla?

**Comment by JackCaoG** [comment]
`we need to add an all_reduce operator to A's backward and B's forward to make the result equivalent to before the splitting` 

You don't need to do that, under the GSPMD you assume there is only one accelerator and you only need to tell compiler how to shard the computation. All of the collective will be inserted by the compiler after SPMD pass.

**Comment by baoleai** [comment]
Thanks. I want to further combine Data Parallel and Tensor Parallel, like TP within a node and DP between nodes, Does XLA support this combination?. 

In DP only case I should sync gradients between nodes like: 
```
output = self.model(input)
loss = output['loss']
loss.backward()
gradients = xm.fetch_gradients(self.optimizer)
xm.dist.all_reduce('sum', gradients, scale=1.0 / xm.xrt_world_size())
```

Should I synchronize gradients across all training processes or just the corresponding local_rank i in each node? For example, if we split A (by row) into two shards on device #0 and device #1, should I only synchronize gradients on device #0 in each node and device #1 in each node, or should I synchronize gradients across all the devices in each node?
